### PR TITLE
Gossip integration test

### PR DESCRIPTION
**Pull request by jframe**
## PR description
Integration tests for gossiping messages to ibft peers.

### CODE REVIEW

File: consensus/ibft/src/integration-test/java/tech/pegasys/pantheon/consensus/ibft/tests/GossipTest.java [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+    final MessageFactory peerMsgFactory = msgCreator.getMessageFactory();
+    final SignedData<ProposalPayload> proposalFromPeer =
+        peerMsgFactory.createSignedProposalPayload(roundId, block);
+
+    gossiper.injectMessage(ProposalMessageData.create(proposalFromPeer));
+
+    final List<ValidatorPeer> validators = new ArrayList<>(roles.getNonProposingPeers());
+    validators.remove(msgCreator);
+    assertPeersContainsReceivedMessages(validators, proposalFromPeer);
+    assertPeersReceivedNoMessages(ImmutableList.of(roles.getProposer(), msgCreator));
```
Comment (CjHare) [inline consensus/ibft/src/integration-test/java/tech/pegasys/pantheon/consensus/ibft/tests/GossipTest.java]: Id suggest having the two assert cases separately. It'll make life easier to debug from Unit failure messages (i.e. without needing to run the code locally)

Comment (jframe) [reply consensus/ibft/src/integration-test/java/tech/pegasys/pantheon/consensus/ibft/tests/GossipTest.java]: The other case where testing that sender is not gossiped to is tested in the other tests as in those tests the sender and the creator are the same. We can't avoid testing that case. What this test is doing differently is testing the case where the sender and the message creator are different nodes and that neither are gossiped too.

File: consensus/ibft/src/integration-test/java/tech/pegasys/pantheon/consensus/ibft/tests/GossipTest.java [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+        msgFactory.createSignedNewRoundPayload(roundId, roundChangeCert, proposal);
+    gossiper.injectNewRound(roundId, roundChangeCert, proposal);
+    assertPeersContainsReceivedMessages(roles.getNonProposingPeers(), newRound);
+
+    gossiper.injectRoundChange(roundId, Optional.empty());
+    assertPeersContainsReceivedMessages(roles.getNonProposingPeers(), roundChange);
+  }
+
+  @Test
+  public void messageIsOnlyGossipedOnce() {
```
[the remark this answers is not in the document]
Comment (jframe) [reply consensus/ibft/src/integration-test/java/tech/pegasys/pantheon/consensus/ibft/tests/GossipTest.java]: done

File: consensus/ibft/src/integration-test/java/tech/pegasys/pantheon/consensus/ibft/support/MessageReceptionHelpers.java [hunk 3]
```diff
@@ -52,15 +53,36 @@ public static void assertPeersReceivedExactly(
             messageMatchesExpected(rxMsgData, msg);
           });
     }
-    allPeers.forEach(p -> p.clearReceivedMessages());
+    allPeers.forEach(ValidatorPeer::clearReceivedMessages);
+  }
+
+  @SafeVarargs
+  public static void assertPeersContainsReceivedMessages(
```
Comment (rain-on) [inline consensus/ibft/src/integration-test/java/tech/pegasys/pantheon/consensus/ibft/support/MessageReceptionHelpers.java]: maybe rename?
assertMessagesReceivedByPeersContains, or
assertPeersReceivedAtLeast?

File: consensus/ibft/src/integration-test/java/tech/pegasys/pantheon/consensus/ibft/support/MessageReceptionHelpers.java [hunk 4]
```diff
@@ -52,15 +53,36 @@ public static void assertPeersReceivedExactly(
...
+    allPeers.forEach(
+        node -> {
+          final List<SignedData<?>> rxMsgs =
+              node.getReceivedMessages()
+                  .stream()
+                  .map(MessageReceptionHelpers::payloadFromMessageData)
+                  .collect(Collectors.toList());
+          assertThat(rxMsgs).containsAll(msgList);
+        });
+    allPeers.forEach(ValidatorPeer::clearReceivedMessages);
```
Comment (rain-on) [inline consensus/ibft/src/integration-test/java/tech/pegasys/pantheon/consensus/ibft/support/MessageReceptionHelpers.java]: nit: SLIGHTLY worried this deletes all the msgs - even ones not checked....but maybe that's fine ...

File: consensus/ibft/src/integration-test/java/tech/pegasys/pantheon/consensus/ibft/support/TestContextBuilder.java [hunk 5]
```diff
@@ -115,6 +115,7 @@ public EventMultiplexer getEventMultiplexer() {
   private IbftEventQueue ibftEventQueue = new IbftEventQueue();
   private int validatorCount = 4;
   private int indexOfFirstLocallyProposedBlock = 0; // Meaning first block is from remote peer.
+  private boolean useGossip;
```
Comment (rain-on) [inline consensus/ibft/src/integration-test/java/tech/pegasys/pantheon/consensus/ibft/support/TestContextBuilder.java]: nit: do we like to explicitly set defaults? (personally I do, but may not be very java-ish)

File: consensus/ibft/src/integration-test/java/tech/pegasys/pantheon/consensus/ibft/support/TestContextBuilder.java [hunk 6]
```diff
@@ -150,9 +156,12 @@ public TestContext build() {
     // Use a stubbed version of the multicaster, to prevent creating PeerConnections etc.
     final StubValidatorMulticaster stubbedMulticaster = new StubValidatorMulticaster();
```
Comment (rain-on) [inline consensus/ibft/src/integration-test/java/tech/pegasys/pantheon/consensus/ibft/support/TestContextBuilder.java]: I think CJH requested I rename this just to multicaster (the stubbed is irrelevant ... could you do it while you're in there?)

File: consensus/ibft/src/integration-test/java/tech/pegasys/pantheon/consensus/ibft/tests/GossipTest.java [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+  private Block block;
+  private ValidatorPeer gossiper;
+  private MessageFactory msgFactory;
+  private SignedData<ProposalPayload> proposal;
+
+  @Before
+  public void setup() {
+    context.getController().start();
+    block = context.createBlockForProposalFromChainHead(roundId.getRoundNumber(), 30);
+    gossiper = roles.getProposer();
```
Comment (rain-on) [inline consensus/ibft/src/integration-test/java/tech/pegasys/pantheon/consensus/ibft/tests/GossipTest.java]: is this node actually the gossiper? I sort of figured the 'unit under test'was the gossiper ...

File: consensus/ibft/src/integration-test/java/tech/pegasys/pantheon/consensus/ibft/tests/GossipTest.java [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+    msgFactory = gossiper.getMessageFactory();
+
+    proposal = msgFactory.createSignedProposalPayload(roundId, block);
+  }
+
+  @Test
+  public void gossipMessagesToPeers() {
+    gossiper.injectProposal(roundId, block);
+    assertPeersContainsReceivedMessages(roles.getNonProposingPeers(), proposal);
```
Comment (rain-on) [inline consensus/ibft/src/integration-test/java/tech/pegasys/pantheon/consensus/ibft/tests/GossipTest.java]: maybe put the negative of this assertion before the injectProposal - to prove the msg didn't used to exist.

nit: And there's probably no harm is saying "all non-proposers contain the prepare msg too"

Do you also need to prove the gossiper DID NOT receive the msg?

File: consensus/ibft/src/integration-test/java/tech/pegasys/pantheon/consensus/ibft/tests/GossipTest.java [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+    block = context.createBlockForProposalFromChainHead(roundId.getRoundNumber(), 30);
+    gossiper = roles.getProposer();
+    msgFactory = gossiper.getMessageFactory();
+
+    proposal = msgFactory.createSignedProposalPayload(roundId, block);
+  }
+
+  @Test
+  public void gossipMessagesToPeers() {
+    gossiper.injectProposal(roundId, block);
```
Comment (rain-on) [inline consensus/ibft/src/integration-test/java/tech/pegasys/pantheon/consensus/ibft/tests/GossipTest.java]: these inject functions actually return the SignedData<?> that they created ... so no need for the message factory

File: consensus/ibft/src/integration-test/java/tech/pegasys/pantheon/consensus/ibft/tests/GossipTest.java (hunk 7 above)
Comment (jframe) [reply consensus/ibft/src/integration-test/java/tech/pegasys/pantheon/consensus/ibft/tests/GossipTest.java]: it's probably more accurately the sender

File: consensus/ibft/src/integration-test/java/tech/pegasys/pantheon/consensus/ibft/support/TestContextBuilder.java (hunk 6 above)
Comment (jframe) [reply consensus/ibft/src/integration-test/java/tech/pegasys/pantheon/consensus/ibft/support/TestContextBuilder.java]: done

File: consensus/ibft/src/integration-test/java/tech/pegasys/pantheon/consensus/ibft/tests/GossipTest.java (hunk 9 above)
Comment (jframe) [reply consensus/ibft/src/integration-test/java/tech/pegasys/pantheon/consensus/ibft/tests/GossipTest.java]: done

File: consensus/ibft/src/integration-test/java/tech/pegasys/pantheon/consensus/ibft/tests/GossipTest.java (hunk 8 above)
Comment (jframe) [reply consensus/ibft/src/integration-test/java/tech/pegasys/pantheon/consensus/ibft/tests/GossipTest.java]: good to have assert before the inject proposal - done
added check that gossiper didn't receive message
also as part of check that gossip didn't receive message confirm it only had the prepare message

File: consensus/ibft/src/integration-test/java/tech/pegasys/pantheon/consensus/ibft/support/MessageReceptionHelpers.java (hunk 4 above)
Comment (jframe) [reply consensus/ibft/src/integration-test/java/tech/pegasys/pantheon/consensus/ibft/support/MessageReceptionHelpers.java]: I think it's fine. I'm expecting that you will use either this message assertion or existing one.

File: consensus/ibft/src/integration-test/java/tech/pegasys/pantheon/consensus/ibft/support/TestContextBuilder.java (hunk 5 above)
Comment (jframe) [reply consensus/ibft/src/integration-test/java/tech/pegasys/pantheon/consensus/ibft/support/TestContextBuilder.java]: Either way, no strong opinion on this one. Slightly less efficient to have an explicit initialisation and could also be considered redundant. However you could argue it's clearer if it was initialised to false.
Often there is a checkstyle or other code style that prevents from having the explicit value as the default.
So i'm happy to initialise to false as we seemed to have followed that convention elsewhere.

File: consensus/ibft/src/integration-test/java/tech/pegasys/pantheon/consensus/ibft/support/MessageReceptionHelpers.java (hunk 3 above)
Comment (jframe) [reply consensus/ibft/src/integration-test/java/tech/pegasys/pantheon/consensus/ibft/support/MessageReceptionHelpers.java]: removed this method as part of refactoring to use existing assert messages method

File: consensus/ibft/src/integration-test/java/tech/pegasys/pantheon/consensus/ibft/tests/GossipTest.java [hunk 10]
```diff
@@ -73,59 +68,56 @@
...
-
-    gossiper.injectCommit(roundId, block.getHash());
-    final IbftExtraData extraData = IbftExtraData.decode(block.getHeader().getExtraData());
-    final Hash commitHash =
-        IbftBlockHashing.calculateDataHashForCommittedSeal(block.getHeader(), extraData);
-    final Signature commitSeal = SECP256K1.sign(commitHash, gossiper.getNodeKeyPair());
-    final SignedData<CommitPayload> commit =
-        msgFactory.createSignedCommitPayload(roundId, block.getHash(), commitSeal);
-    assertPeersContainsReceivedMessages(roles.getNonProposingPeers(), commit);
+    final SignedData<PreparePayload> prepare = gossiper.injectPrepare(roundId, block.getHash());
```
Comment (rain-on) [inline consensus/ibft/src/integration-test/java/tech/pegasys/pantheon/consensus/ibft/tests/GossipTest.java]: technically the gossiper SHOULDN'T send in the prepare (proposer can't also prepare) ... but not too worried.

File: consensus/ibft/src/integration-test/java/tech/pegasys/pantheon/consensus/ibft/tests/GossipTest.java (hunk 7 above)
Comment (jframe) [reply consensus/ibft/src/integration-test/java/tech/pegasys/pantheon/consensus/ibft/tests/GossipTest.java]: renamed to sender

File: consensus/ibft/src/integration-test/java/tech/pegasys/pantheon/consensus/ibft/tests/GossipTest.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (jframe) [reply consensus/ibft/src/integration-test/java/tech/pegasys/pantheon/consensus/ibft/tests/GossipTest.java]: it's really done this time. change wasn't included in the git push unfortunately

**Comment by jframe** [comment]
> Do you have any tests where a msg is signed by one Peer, but injected by another? Then make sure it didn't go to either of them?

There is a test for this, it's called "messageIsNotGossipedToSenderOrCreator"

File: consensus/ibft/src/integration-test/java/tech/pegasys/pantheon/consensus/ibft/tests/GossipTest.java (hunk 1 above)
Comment (CjHare) [reply consensus/ibft/src/integration-test/java/tech/pegasys/pantheon/consensus/ibft/tests/GossipTest.java]: Right, I understand ...my issue is actually with the underlying function.

**Pull request merged**